pkg/verify/verify.go

```diff
@@ -0,0 +1,204 @@
+//
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package verify
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/cyberphone/json-canonicalization/go/src/webpki.org/jsoncanonicalizer"
+	"github.com/sigstore/rekor/pkg/generated/client"
+	"github.com/sigstore/rekor/pkg/generated/client/tlog"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/util"
+	"github.com/sigstore/sigstore/pkg/signature"
+	"github.com/sigstore/sigstore/pkg/signature/options"
+	"github.com/transparency-dev/merkle/proof"
+	"github.com/transparency-dev/merkle/rfc6962"
+)
+
+// ProveConsistency verifies consistency between an initial, trusted STH
+// and a second new STH. Callers MUST verify signature on the STHs'.
+func ProveConsistency(ctx context.Context, rClient *client.Rekor,
+	oldSTH *util.SignedCheckpoint, newSTH *util.SignedCheckpoint, treeID string) error {
+	oldTreeSize := int64(oldSTH.Size)
+	switch {
+	case oldTreeSize == int64(newSTH.Size):
+		if !bytes.Equal(oldSTH.Hash, newSTH.Hash) {
+			return errors.New("old root hash does not match STH hash")
+		}
+	case oldTreeSize < int64(newSTH.Size):
+		consistencyParams := tlog.NewGetLogProofParamsWithContext(ctx)
+		consistencyParams.FirstSize = &oldTreeSize      // Root size at the old, or trusted state.
+		consistencyParams.LastSize = int64(newSTH.Size) // Root size at the new state to verify against.
+		consistencyParams.TreeID = &treeID
+		consistencyProof, err := rClient.Tlog.GetLogProof(consistencyParams)
+		if err != nil {
+			return err
+		}
+		var hashes [][]byte
+		for _, h := range consistencyProof.Payload.Hashes {
+			b, err := hex.DecodeString(h)
+			if err != nil {
+				return errors.New("error decoding consistency proof hashes")
+			}
+			hashes = append(hashes, b)
+		}
+		if err := proof.VerifyConsistency(rfc6962.DefaultHasher,
+			oldSTH.Size, newSTH.Size, hashes, oldSTH.Hash, newSTH.Hash); err != nil {
+			return err
+		}
+	case oldTreeSize > int64(newSTH.Size):
+		return errors.New("inclusion proof returned a tree size larger than the verified tree size")
+	}
+	return nil
+
+}
+
+// VerifyCurrentCheckpoint verifies the provided checkpoint by verifying consistency
+// against a newly fetched Checkpoint.
+//nolint
+func VerifyCurrentCheckpoint(ctx context.Context, rClient *client.Rekor, verifier signature.Verifier,
```

Comment:
It seems like this func assumes the we are verifying against the active shard, so can we omit the `treeID` argument? I think we can grab it from `logInfo := result.GetPayload()` later in the function

Reply:
Nice catch!